File: src/in_memory_database_migration/javascript/DatabaseMigration.js
const { Pool } = require('pg');

class DatabaseMigration {
  constructor(source, target, batchSize = 1000) {
    this.source = source; 
    this.target = target;
    this.batchSize = batchSize;
    this.sourcePool = new Pool(this.source);
    this.targetPool = new Pool(this.target);
  }

  async extract() {

    let data = [];
    const client = await this.sourcePool.connect();
    try { 
      const result = await client.query('SELECT * FROM source_table');
      data = result.rows; 
    } finally {
      client.release()
    }
    return data; 
  }

  transform(data) {
    // Placeholder for any data transformations
    return data;
  }

  async load(data) {
    let index = 0;
    while (index < data.length) {
      const batch = data.slice(index, index + this.batchSize);
      const values = batch.map((row) => `(${row.id}, '${row.name}')`).join(',');
      const client = await this.targetPool.connect();
      try {
        await client.query(`INSERT INTO target_table (id, name) VALUES ${values}`);
      } finally {
        client.release();
      }
      index += this.batchSize;
    }
  }

  async perform() {
    const data = this.extract(); 
    const transformedData = this.transform(data);
    this.load(transformedData);
  }
}

// Example usage: 

const source = {
  user: process.env.SOURCE_DB_USER,
  host: process.env.SOURCE_DB_HOST,
  database: process.env.SOURCE_DB_NAME,
  password: process.env.SOURCE_DB_PASSWORD,
  port: process.env.SOURCE_DB_PORT,
};

const target = {
  user: process.env.TARGET_DB_USER,
  host: process.env.TARGET_DB_HOST,
  database: process.env.TARGET_DB_NAME,
  password: process.env.TARGET_DB_PASSWORD,
  port: process.env.TARGET_DB_PORT,
};

const batchSize = process.env.BATCH_SIZE || 1000;

const etl = new DatabaseMigration(source, target, batchSize);
etl.perform();

File: src/in_memory_database_migration/go/DatabaseMigration.go
/*
This script uses the Go database/sql package to connect to a source and destination database, extract data from the source database, transform the data, and load the transformed data into the destination database.

Note that this is a simple example, and a real-world ETL script would likely be more complex and include additional error handling and logging.
*/

package main

import (
    "database/sql"
    "fmt"
    _ "github.com/go-sql-driver/mysql"
)

func main() {
    // Connect to the source database
    sourceDB, err := sql.Open("mysql", "user:password@tcp(localhost:3306)/source_db")
    if err != nil {
        panic(err)
    }
    defer sourceDB.Close()

    // Connect to the destination database
    destDB, err := sql.Open("mysql", "user:password@tcp(localhost:3306)/dest_db")
    if err != nil {
        panic(err)
    }
    defer destDB.Close()

    // Extract data from the source database
    rows, err := sourceDB.Query("SELECT id, name, email FROM users")
    if err != nil {
        panic(err)
    }
    defer rows.Close()

    // Transform the data
    var id int
    var name string
    var email string
    for rows.Next() {
        err := rows.Scan(&id, &name, &email)
        if err != nil {
            panic(err)
        }

        // Example transformation: concatenate name and email
        transformedName := fmt.Sprintf("%s <%s>", name, email)

        // Load the transformed data into the destination database
        _, err = destDB.Exec("INSERT INTO users (id, name) VALUES (?, ?)", id, transformedName)
        if err != nil {
            panic(err)
        }
    }
}
